tools/codegen/gen.py

```diff
@@ -231,41 +231,80 @@ def get_static_dispatch_backend(f: NativeFunction, backend_index: BackendIndex)
+    cpp_sig: CppSignature, *,
+    method: bool,
+    backend_index: Optional[BackendIndex]
+) -> str:
+    if backend_index is None or f.manual_kernel_registration:
+        return None
+    target_sig = CppSignatureGroup.from_native_function(f, method=False, fallback_binding=False).signature
+    name = target_sig.name()
+    exprs = translate(cpp_sig.arguments(), target_sig.arguments(), method=method)
+    exprs_str = ', '.join(a.expr for a in exprs)
+    if f.structured_delegate is not None:
+        # TODO: for ops with structured_delegate it should check the dispatch table of
+        # the out variant instead. For now, these structured ops all have CPU/CUDA kernels
+        # so we always dispatch to the `backend`, but this could be wrong when we
+        # migrate math/default_backend ops to use structured delegate.
+        if backend_index.has_kernel(f) or backend_index.dispatch_key in STRUCTURED_DISPATCH_KEYS:
+            return f'return at::{backend_index.dispatch_key.lower()}::{name}({exprs_str});'
+        else:
+            return f'TORCH_CHECK(false, "Static dispatch does not support {name} for {backend_index.dispatch_key}.");'
 
+    if backend_index.has_kernel(f):
+        return f'return at::{backend_index.dispatch_key.lower()}::{name}({exprs_str});'
+    elif f.has_composite_explicit_autograd_kernel:
+        return f'return at::{DispatchKey.CompositeExplicitAutograd.lower()}::{name}({exprs_str});'
+    elif f.has_composite_implicit_autograd_kernel:
+        return f'return at::{DispatchKey.CompositeImplicitAutograd.lower()}::{name}({exprs_str});'
+
+    return f'TORCH_CHECK(false, "Static dispatch does not support {name} for {backend_index.dispatch_key}.");'
 
 def static_dispatch(
         f: NativeFunction, cpp_sig: CppSignature,
-        *, method: bool, backend_index: Optional[BackendIndex]
+        *, method: bool, backend_indices: List[BackendIndex]
 ) -> Optional[str]:
-    if backend_index is None or f.manual_kernel_registration:
+    if backend_indices is None or f.manual_kernel_registration:
         return None
     target_sig = CppSignatureGroup.from_native_function(f, method=False, fallback_binding=False).signature
     name = target_sig.name()
     exprs = translate(cpp_sig.arguments(), target_sig.arguments(), method=method)
     exprs_str = ', '.join(a.expr for a in exprs)
 
-    dispatch_key = get_static_dispatch_backend(f, backend_index)
-    if dispatch_key is not None:
-        return f'return at::{dispatch_key.lower()}::{name}({exprs_str});'
-
-    return f'TORCH_CHECK(false, "Static dispatch does not support {name} for {backend_index.dispatch_key}.");'
+    keys = [b for b in backend_indices if b.has_kernel(f) or f.structured_delegate is not None]
+    if len(keys) == 1:
+        return generate_static_dispatch(f, cpp_sig, method=method, backend_index=keys[0])
+    elif len(keys) == 0:
+        return generate_static_dispatch(f, cpp_sig, method=method, backend_index=backend_indices[0])
```

Comment:
What's the idea behind passing `backend_indices[0]` here?

Reply:
@larryliu0820 
If none of the static dispatch backends have kernels for the native function, then we just pass in the first one in the list, could be any one of them, as in generate_static_dispatch(f, ..) if the backend doesn't have a kernel or if the native function doesn't have a structured delegate we only see if the native function has compositeImplicitAugradKernel or compositeImplicitAugradKernel and use that if not emit a TORCH_CHECK https://github.com/pytorch/pytorch/pull/75605/files#diff-6382c847dac2d011f822c492e61f9fa0eceb47c2d6fc5fb4f0c42af726c1a3aaR283-R288